libraries/chain/backing_store/tests/benchmark_kv.cpp

```diff
@@ -0,0 +1,337 @@
+      exit(2);
+   }
+
+   std::string key;
+   auto i = 0;
+   std::unordered_set<std::string> unordered_keys;
+   while (i < work_set_size && getline(key_file, key)) {
+      unordered_keys.insert(key);  // randomize
+      ++i;
+   }
+   key_file.close();
+
+   std::vector<std::string> keys;
+   for (auto& k: unordered_keys) {
+      keys.push_back(k);
+   }
+   return keys;
+}
+
+void create_key_values(cmd_args& args, const std::unique_ptr<kv_context>& kv_context_ptr) {
+   std::ifstream key_file(args.key_file);
+
+   if (!key_file.is_open()) {
+      std::cout << "Failed to open key file " << args.key_file << std::endl;
+      exit(2);
+   }
+
+   std::string value(args.value_size, 'a');
+   std::string key;
+   while (getline(key_file, key)) {
+      kv_context_ptr->kv_set(contract, key.c_str(), key.size(), value.c_str(), value.size(), payer);
+      ++args.num_keys;
+   }
+
+   key_file.close();
+}
+
+void benchmark_get(cmd_args& args, const std::unique_ptr<kv_context> kv_context_ptr, std::vector<std::string> work_set, wall_clock_t& start, wall_clock_t& end, rusage& usage_start, rusage& usage_end) {
+   start = std::chrono::steady_clock::now();
+   getrusage(RUSAGE_SELF, &usage_start);
+
+   for (auto i = 0; i < num_iterations; ++i) {
+      for (auto& key: work_set) {
+         uint32_t actual_value_size;
+         kv_context_ptr->kv_get(contract, key.c_str(), key.size(),actual_value_size);
+      }
+   }
+
+   getrusage(RUSAGE_SELF, &usage_end);
+   end = std::chrono::steady_clock::now();
+}
+
+void benchmark_get_data(cmd_args& args, const std::unique_ptr<kv_context> kv_context_ptr, std::vector<std::string> work_set, wall_clock_t& start, wall_clock_t& end, rusage& usage_start, rusage& usage_end) {
+   char* data = new char[args.value_size];
+
+   start = std::chrono::steady_clock::now();
+   getrusage(RUSAGE_SELF, &usage_start);
+
+   for (auto i = 0; i < num_iterations; ++i) {
+      for (auto& key: work_set) {
+         kv_context_ptr->kv_get_data(0, data, default_value_size);
+      }
+   }
+
+   getrusage(RUSAGE_SELF, &usage_end);
+   end = std::chrono::steady_clock::now();
+}
+
+void benchmark_set(cmd_args& args, const std::unique_ptr<kv_context> kv_context_ptr, std::vector<std::string> work_set, wall_clock_t& start, wall_clock_t& end, rusage& usage_start, rusage& usage_end) {
+   std::string value(args.value_size, 'b');
+
+   start = std::chrono::steady_clock::now();
+   getrusage(RUSAGE_SELF, &usage_start);
+
+   for (auto i = 0; i < num_iterations; ++i) {
+      for (auto& key: work_set) {
+         kv_context_ptr->kv_set(contract, key.c_str(), key.size(), value.c_str(), value.size(), payer);
+      }
+   }
+
+   getrusage(RUSAGE_SELF, &usage_end);
+   end = std::chrono::steady_clock::now();
+}
+
+void benchmark_it_next(cmd_args& args, const std::unique_ptr<kv_context> kv_context_ptr, wall_clock_t& start, wall_clock_t& end, rusage& usage_start, rusage& usage_end) {
```

Comment:
It seems like we would also want to know the overhead of creating an iterator.  So another function like this but just create random iterators from our key set.

Reply:
Will add that. Thanks you Bucky and Steven for the excellent comments!